### DISCUSSION TITLE
📊Serverless Databases Performance Comparison

### DISCUSSION THREAD

**Discussion by dajinchu** [blitz-js/blitz#503] [General]
Would love to start some discourse on how to get great performance out of a serverless deployment and how database choice factors into that. This is in the hopes of guiding our users towards the pit of success with a "performant by default" solution.

## An Experiment

I'm going to test out a bunch of different configurations and get some numbers. Note that this experiment is more about nextjs/prisma/serverless than Blitz 🤷 I couldn't find any existing info on this, so here we are. Maybe the Vercel people know more.

Controlled variables: 
- Running app from `blitz new`
- Using same data model as Blitz Store example
- Testing custom api route to just hit `db.product.findMany({ first: 4 })` and 500 if error is thrown.
- Just 4 products in the DB. This should be a fast operation, looking more into network bottlenecks than postgres db query performance
- Using Artillery.io to send 20 requests/second for 30 seconds

I will post about each database setup in a reply here.

Repo: https://github.com/dajinchu/blitz-perftest

## Keeping Score
![image](https://user-images.githubusercontent.com/7122182/81841945-2ecb0700-9519-11ea-8876-2c16bce66ff3.png)

**Comment by dajinchu** [comment]
# 1. Localhost + DigitalOcean

Blitz started in production mode.

laptop in Boston. Digital Ocean in NYC1

| Blitz | DB | DB Host | DB Cost | DB stats |
| -----| ----| ----------| --------  | ----------- |
| localhost | Postgres | DigitalOcean | $15/month |  1 GB RAM / 1vCPU / 10 GB Disk |

Results
```
  Response time (msec):
    min: 24.1
    max: 52
    median: 27.3
    p95: 32.6
    p99: 42
  Codes:
    200: 600
```

**Comment by dajinchu** [comment]
# 2. Vercel + Digital Ocean

Vercel in aws us-east-1, digitalocean in nyc1

| Blitz | DB | DB Host | DB Cost | DB stats |
| -----| ----| ----------| --------  | ----------- |
| Vercel | Postgres | DigitalOcean | $15/month |  1 GB RAM / 1vCPU / 10 GB Disk |

```
  Response time (msec):
    min: 108.8
    max: 1213.5
    median: 125.1
    p95: 145.6
    p99: 293.7
  Codes:
    200: 600
```

With a bit more load or time, we start seeing 500's with the error
`FATAL: remaining connection slots are reserved for non-replication superuser connections`

This is because for serverless deploys, each request can make another database connection, so the 22 database connections allotted by DO quickly runs out. This should be solved by proxying the database with a layer that manages the connection pooling. 

Relevant prisma docs: https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client/deployment#serverless-faas

## w/DigitalOcean's Connection Pool (pgBouncer)

DigitalOcean can setup pgBouncer to do connection pooling for us. They say that they can support 5000 connections per pool this way. Great!

With pooling:
```
  Response time (msec):
    min: 128
    max: 1310.8
    median: 148.4
    p95: 194.1
    p99: 800.4
  Codes:
    200: 600

```
There used to errors with some variant of: `"SqlState xx, prepared statement 'sxx' does not exist."`

Closed in https://github.com/prisma/prisma/issues/2520

## Conclusion
DigitalOcean Postgres now works well with Prisma, just be sure to enable the connection pool on DigitalOcean!

**Reply by flybayer** [reply]
I think we should remove the Vercel + DB without PG bouncer case (and remove from the graphic above), because it's just a fundamental aspect of serverless that you have to have a separate connection pool for traditional DBs

**Reply by dajinchu** [reply]
agreed

**Reply by flybayer** [reply]
Also, let's add more details about DB location plus vercel region

**Reply by dajinchu** [reply]
yep, that is in the works. redeploying and testing to get new stats. I do think there is some educational value in showing the importance of proper configuration w/region+pooling (and help our users get setup with them). Though as far as comparing the performance of different providers, I agree we should just show the most optimal setup for each.

**Reply by dajinchu** [reply]
perhaps in some tutorials elsewhere we can show how critical flipping a few settings is

**Reply by flybayer** [reply]
Ok yeah, good point on having a reference for what the connection pool does

**Reply by janpio** [reply]
If you have any idea how we at @prisma can improve pgBouncer support with DO, let us know: Using `forceTransactions: true` is the right way to go for pgBouncer as you found out. ~~Unfortunately we did not find a good way to configure pgBouncer correctly on DO. But we will take another look and see if we missed anything.~~ In theory it should just work ™️ (We do have an e2e test that is run for each Prisma version that makes sure it still works, even explicitly with DO Postgres pgBouncer!) but seems you are getting a few 500 there - we will look into this and provide and update here.

**Reply by janpio** [reply]
We can reproduce some errors with Prisma, `forceTransactions:true` and DigitalOcean's PgBouncer (Connection Pooling) and are tracking our further investigation and work here: https://github.com/prisma/prisma/issues/1638#issuecomment-628235096 I'm pretty confident we can provide a fix pretty soon - will keep you updated.

**Comment by dajinchu** [comment]
# 3. Just DigitalOcean

Ran Blitz on a DigitalOcean droplet + connected to DigitalOcean DB from setup 1 and 2. Everything in digitalocean nyc1

| Blitz | DB | DB Host | DB Cost | DB stats |
| -----| ----| ----------| --------  | ----------- |
| DigitalOcean: 1 GB Memory / 25 GB Disk ($5/month) | Postgres | DigitalOcean | $15/month |  1 GB RAM / 1vCPU / 10 GB Disk |

Results:
```
  Response time (msec):
    min: 25.8
    max: 1094.7
    median: 33.2
    p95: 82.2
    p99: 110.6
  Codes:
    200: 600
```

This is an **4x faster** than Vercel, and competitive with setup 1. It's not a fair test exactly because both resources are in DigitalOcean, but theoretically, the Vercel setup should be close by (DO NYC1 to aws iad1). A more fair test will be conducted with RDS in the exact same datacenter as Vercel. Stay tuned :)

**Comment by matheuss** [comment]
Hey! Matheus from Vercel here. I found a few misleading points on your analysis:

Your Vercel deployment is far away from your database. As per your GitHub profile, I see you live in Boston, and given you got good results from localhost, I assume you're using the NYC DO region. Your Vercel deployment, however, is not configured to be deployed any specific region (no `regions` [here](https://github.com/dajinchu/blitz-perftest/blob/45627cd33f1cc58f20b58c920ec64d0a66c2a4d5/now.json)), which happens to default to `sfo1` for deployments made with [Vercel for GitHub](https://vercel.com/github). 

Therefore, your results with Vercel are misleading (or, arguably, invalid): **each database query results in a roundtrip across the US**! I've opened a PR on your repo with a fix: https://github.com/dajinchu/blitz-perftest/pull/1 👍 

---

Another point you mentioned:

>You can see why Vercel is so slow - we have to send packets over the public internet twice.

The networking between cloud providers is known to be optimised – it's on their best interest to have an optimal connection to other providers. I have a VM on DO NYC1, and you can see that the **latency to the nearest Vercel Edge** (`iad1`) is only **half a millisecond**:

```
root@do-vm-nyc:~# ping -c 4 vercel.com
PING vercel.com (76.76.21.21) 56(84) bytes of data.
64 bytes from 76.76.21.21: icmp_seq=1 ttl=121 time=0.474 ms
64 bytes from 76.76.21.21: icmp_seq=2 ttl=121 time=0.503 ms
64 bytes from 76.76.21.21: icmp_seq=3 ttl=121 time=0.534 ms
64 bytes from 76.76.21.21: icmp_seq=4 ttl=121 time=0.513 ms

--- 76.76.21.21 ping statistics ---
4 packets transmitted, 4 received, 0% packet loss, time 3049ms
rtt min/avg/max/mdev = 0.474/0.506/0.534/0.021 ms
```
That's less than the latency from my laptop to my Wi-Fi router 😄 

---

Furthermore, this is also misleading:

>20 requests/sec is not exceptionally high

If I understand [their docs](https://artillery.io/docs/cli-reference/) correctly, you're telling Artillery to do 20 req/s (`-r 20`), so 20 req/s is to be expected.

---

Regarding the high rate of errors, that seems to be a limitation of Digital Ocean, given you mentioned that it's not possible to apply the fix that Prisma suggested.

**Reply by dajinchu** [reply]
Awesome, thanks for figuring that out. I'll test it out with the region fix. I was suspicious of the high latency - especially since I thought Vercel ran on AWS anyway :)

**Reply by dajinchu** [reply]
On the point about the requests/sec, I meant it as, "we shouldn't be overwhelmed at only 20 req/sec." Will edit as I understand that it is not clear.

**Reply by matheuss** [reply]
But that's a limitation of the database, right? You're limited to 20 database connections, so you can't have more than 20 concurrent requests if you do a query on each request.

**Reply by dajinchu** [reply]
Yes, totally. Vercel with RDS is doing a lot better, thanks to higher connection limits. However, it is not an issue that regular server deployments face since they won't initiate new connections on each request.

**Reply by flybayer** [reply]
**Everything is now updated to show stats with Vercel region same as the Digital Ocean DB region.**

But even with this we're still at ~120ms latency with Vercel + DO. Will try to investigate this more.

**Comment by dajinchu** [comment]
# 4. AWS EC2 + RDS

Baseline setup with RDS. No RDS proxy running. Equivalent to setup 3. Everything in us-east-1.

| Blitz | DB | DB Host | DB Cost | DB stats |
| -----| ----| ----------| --------  | ----------- |
| EC2 t2-micro (1vCPU, 1GB RAM, ~$8/month) | Postgres | RDS | ~$15/month |  1 vCPU, 1 GB RAM, 20GB disk |
```
  Response time (msec):
    min: 41
    max: 1131.7
    median: 45.6
    p95: 55.4
    p99: 154.4
  Codes:
    200: 600
```

**Comment by dajinchu** [comment]
# 5. Vercel + RDS

Vercel and RDS both in us-east-1. 

```
  Response time (msec):
    min: 94.7
    max: 1205.7
    median: 109.7
    p95: 153.7
    p99: 714.5
  Codes:
    200: 600
```

Just about the same as setup 2. The difference could be due to being in the same datacenter instead of 2 nearby datacenters, but there are other uncontrolled variables like switching db from DO to RDS. Since we know that Vercel and RDS are in the same datacenter, I wonder why we're still 50ms slower than the EC2 solution. Could just be the time it takes to go through Vercel's load balancing/routing infrastructure?

With enough load, we're able to 500 the server due to running out of connections, so we need to setup a connection pooling solution. Unfortunately, RDS Proxy can only connect within the same VPC, afaik. Will continue digging.

**Comment by flybayer** [comment]
@dajinchu question: is the `max` stat for initial DB connection? Or is the DB already connected before you start the test?

Couple other things I'd also like to know:

- How much latency is Prisma adding?
- What is the latency for using RDS from vanilla Lambda?

**Reply by dajinchu** [reply]
It does not include initial connection. I turned on Prisma logging and made sure we connected once first.

Will investigate the other two over the weekend!

**Reply by dajinchu** [reply]
Prisma appears to be negligible. Wrote a raw query using `pg` with the same SQL query that prisma sends. (see here https://github.com/dajinchu/blitz-perftest/blob/master/app/api/raw.ts#L10-L11). Running this on Vercel+AWS RDS, Prisma gets median response time 110ms, pg gets 107ms

**Reply by dajinchu** [reply]
Not a fair test because I just loaded up a dead simple lambda, as opposed to all of next/blitz, but running the pg query in a lambda gives us a median response time of 94ms. Vercel is only 15ms slower, and it could totally be chalked up to the difference in the function that is tested. I'll test out putting actual next.js code on a serverless function to see if there's any difference, but it feels like Vercel and AWS Lambda are competitive. Server deployments still are in another league though.

**Reply by dajinchu** [reply]
Next move for me might be checking on ECS, though I'm guessing it'll be equivalent to EC2. Also investigating what exactly is making EC2 2x faster. (is it all lambda startup time?)
